changelog/dip1000_deprecation_warnings.dd

```diff
@@ -1,6 +1,45 @@
-Print warning messages unless -revert=dip1000 is used
+Print deprecations for `scope` pointer errors
 
-As an incentive to transition to -preview=dip1000 by default, now the
-compiler prints what would have been error messages if the flag had
-been on as deprecation messages.  This new behaviour can be disabled
-with -revert=dip1000.
+The `scope` attribute has existed for a long time, but the compiler would only verify its semantics when the `-preview=dip1000` switch was passed, to avoid breaking code.
+A variable marked `scope` is assumed to free the memory it points to when it goes out of scope, so `scope` variables can only be assigned to variables that will be destructed before itself.
```

Comment:
This seems actively misleading to me. `scope` is about escape analysis. `scope` variables do not free memory, and the lifetime analysis is based on lexical scoping, not the order of destructor calls (e.g., you can `destroy` the target of a `scope` pointer before the pointer goes out of scope).

Reply:
The lifetime analysis is based on lexical scoping because that's what the order of destructor calls is based on, and destructing a pointer early is/should be `@system`. I agree it is misleading to say "is assumed to free the memory", I tried to describe them without referring to 'lifetimes' because there's no clear definition of that, but I can resort to expanding the text a bit.

Reply:
You could maybe phrase it in terms of "escaping"; something like

> Pointers or references stored in a `scope` variable are not allowed to escape the scope in which the variable is defined.

...which IMO is the core concept. But given that the compiler messages users will be seeing are all phrased in terms of lifetimes ("`x` assigned to `y` with longer lifetime"), I don't think it makes sense to try to avoid that term here.

Reply:
> destructing a pointer early is/should be `@system`

Given a pointer `p`:

* `destroy(p)` is `@safe`—it sets `p` to `null`.
* `destroy(*p)` is `@safe` as long as `*p`'s destructor is `@safe`, and can happen even if `*p` has a `scope` pointer pointing to it (indeed, even if `p` is itself a `scope` pointer).